packages/Config/RectorConfig.php

```diff
@@ -163,14 +168,28 @@ public function ruleWithConfiguration(string $rectorClass, array $configuration)
         Assert::isAOf($rectorClass, RectorInterface::class);
         Assert::isAOf($rectorClass, ConfigurableRectorInterface::class);
 
-        $this->singleton($rectorClass);
-        $this->afterResolving($rectorClass, static function (ConfigurableRectorInterface $configurableRector) use (
+        // store configuration to cache
+        $this->ruleConfigurations[$rectorClass] = array_merge(
+            $this->ruleConfigurations[$rectorClass] ?? [],
             $configuration
-        ): void {
-            $configurableRector->configure($configuration);
-        });
+        );
+
+        $isBound = $this->bound($rectorClass);
+
+        // avoid double registration
+        if ($isBound) {
+            return;
+        }
 
+        $this->singleton($rectorClass);
         $this->tagRectorService($rectorClass);
+
+        $this->afterResolving($rectorClass, function (ConfigurableRectorInterface $configurableRector) use (
+            $rectorClass
+        ): void {
+            $ruleConfiguration = $this->ruleConfigurations[$rectorClass];
+            $configurableRector->configure($ruleConfiguration);
+        });
```

Comment:
The `configure()` is now called exactly once with the full configuration.

We might check for the configuration override and have custom property to add this after resolving too. In case of non-configurable rules, but it might be too defensive. Let's see how it performs :)